include/zephyr/bluetooth/gatt.h

```diff
@@ -55,14 +55,18 @@ enum {
 	/** @brief Attribute read permission with authentication.
 	 *
 	 *  If set, requires encryption using authenticated link-key for read
-	 *  access.
+	 *  access. This further means that if set BT_GATT_PERM_READ_ENCRYPT is
+	 *  implied, and can be omitted. It does however not imply
+	 *  BT_GATT_PERM_READ_AUTHEN.
 	 */
 	BT_GATT_PERM_READ_AUTHEN = BIT(4),
 
 	/** @brief Attribute write permission with authentication.
 	 *
 	 *  If set, requires encryption using authenticated link-key for write
-	 *  access.
+	 *  access. This further means that if set BT_GATT_PERM_WRITE_ENCRYPT is
+	 *  implied, and can be omitted. It does however not imply
+	 *  BT_GATT_PERM_WRITE_AUTHEN.
```

Comment:
Regarding the OOB, my understanding do is that the OOB can yield key sizes all the way from 7-16 octets, as it can be used for e.g. Legacy Pairing too.

My question then is, as this sort of feel likes another extension, if this functionality better fit in another issue?

Reply:
Also, BT_GATT_PERM_WRITE_AUTHEN as you have suggested to remove was an error, it was supposed to say 128B_ENTROPY. I dont know if that makes you still dislike it.

Reply:
`128B_ENTROPY` sounds like it makes sense :)

My understanding is also that `key_size != entropy`; entropy size of 128-bits means that the data used to generate the key was at least 128-bit, and that you can technically generate a 56-bit key with e.g. 256-bits of entropy, and vice versa.

Reply:
That does make sense, but you are still not sure in regards to that entropy level being used for OOB I assume?

Reply:
Correct. My theory is that OOB *can* provide 128-bits of entropy, but perhaps does not ensure it (unlike LE Secure Connections), but I cannot find anything in the core spec that specifically states any requirements for the OOB procedure that will ensure 128-bits of entropy.

Reply:
> But a 65-bit key certainly cannot have 256-bits of entropy in my mind.

Entropy is the randomness that is used to generate a key, and can be as large as one wants it to be. So you can generate a 8-bit key with 1028-bit entropy if you want. You can also always truncate a key without losing entropy, right? So I don't think there's any hard relationsship between key size and entropy. 

We know that LESC have 128-bit entropy because of the diffie hellman algorithm used, but for OOB method we don't really know what kind of entropy is used, do we? Or does the BT SMP ensure that at least 128-bit entropy is used?

Reply:
> Correct. My theory is that OOB _can_ provide 128-bits of entropy, but perhaps does not ensure it (unlike LE Secure Connections), but I cannot find anything in the core spec that specifically states any requirements for the OOB procedure that will ensure 128-bits of entropy.

Legacy OOB exchanges the TK. The whole pairing process is protected against MITM only if the TK has sufficient entropy. Core spec 5.3 Vol 3 Part H 2.3.5.4 says:

> The TK value shall be a 128-bit random number using the requirements for random generation defined in [Vol 2] Part H, Section 2.

Reply:
OK, so it sounds like OOB will always provide 128-bit entropy, assuming that the random value generated complies with the requirement from the core spec. 

In that case we can use the 128B_ENTROPY and then simply do a check with `LESC or OOB` :) Thanks

@larsgk @fredrikdanebjer Do you plan on extending this PR to also cover the OOB case?

Reply:
@Thalley No, we would actually prefer on not doing it at this point. Can we maybe create an issue for it for future work?

Reply:
> You can also always truncate a key without losing entropy, right?

I disagree with this. The entropy is accounting for the weakest link in the chain. The key size limits the entropy.

Consider a key size of n-bit with each bit independently sampled from a random uniform distribution with elements {0, 1}. That results in n bits of entropy. There is no way to cram more randomness/uncertainty/entropy into the key. It can be brute-forced in 2^n attempts, on average half that. A 1 bit key can be brute-forced in two attempts, always. The key has no memory of the amount of entropy that were wasted and did not end up in the key during the generation.

> Entropy is the randomness that is used to generate a key, and can be as large as one wants it to be. So you can generate a 8-bit key with 1028-bit entropy if you want.

There is some sense to this when considering the entropy of the along with its source if that is a attack vector. A key made from two 128-bit keys xor-ed together still only has 128 bits of entropy. If one of the two key original keys are leaked, the result key still has 128 bits of entropy. Only when both keys are leaked, the result key has 0 entropy in this system. So you can leak 128 bits of information about the input to the xor without compromising the key. So in some stretch of the imagination, you can leak 128 bits, which are in some way about the result key, without compromising it. This may be relevant if the entropy source is lying about it's entropy or there is some other leak. E.g. if an attacker can eavesdrop some of the randomness it produces. Then by using more randomness from that source, you can get the required key entropy after all.

Reply:
> @Thalley No, we would actually prefer on not doing it at this point. Can we maybe create an issue for it for future work?

If we add a bit for 128B entropy and omit the OOB check, then that needs to be documented, and we should create a issue for it right away IMO :)

Reply:
> I disagree with this. The entropy is accounting for the weakest link in the chain. The key size limits the entropy.

You are right :)